### ISSUE TITLE
Agda: Theme does not impact syntax highlighting

### ISSUE THREAD

**Issue by hydrobeam** [protesilaos/modus-themes#48]
Hi, I first just wanted to thank you for the excellent themes!

When using agda, you can load a file (with `C-x C-c C-l`) which colours the buffer, however when using the modus-themes (and ef-themes), the colours don't seem to change from their "default" colours, which impacts visibility.

However, I noticed that while experimenting with [`doom-themes`](https://github.com/hlissner/emacs-doom-themes), the highlighting corresponds to the theme used. I'm not sure if this is because of an upstream`agda` issue or a `modus` thing though.

I found this issue in the agda repo which seems to imply it might be an agda thing: https://github.com/agda/agda/issues/5917

> Addressing the syntax highlighting situation. The current approach has difficulties with certain themes, making it difficult to read or recognize some of the things that Agda wishes to point out. To this end I'd like to investigate if setting agda2-highlight-face-groups to default-faces is feasible.

Here are some screenshots and comparisons:

### `modus-vivendi`
<img width="778" alt="image" src="https://user-images.githubusercontent.com/70682032/199151902-bdee39f5-f7a7-4c0a-8f3e-14592d467df5.png">

### `modus-operandi`
<img width="740" alt="image" src="https://user-images.githubusercontent.com/70682032/199151949-15f7ee72-9e03-4a9e-810a-b83f8ef17b5f.png">

### `ef-bio`
<img width="711" alt="image" src="https://user-images.githubusercontent.com/70682032/199153993-10f7a6f1-604f-49eb-a8af-72b28faa304a.png">

### `ef-tritanopia-dark`
<img width="677" alt="image" src="https://user-images.githubusercontent.com/70682032/199152054-0a15ea79-04a6-4045-9e56-9f45be4ad9f4.png">

### `doom-vibrant`
<img width="667" alt="image" src="https://user-images.githubusercontent.com/70682032/199152412-222a19fe-1146-4bd2-b6d8-9c99880dde69.png">

### `doom-moonlight`
<img width="684" alt="image" src="https://user-images.githubusercontent.com/70682032/199152447-5690fdf4-84c3-4441-b663-5c138ce9c64a.png">

### `doom-one`
<img width="708" alt="image" src="https://user-images.githubusercontent.com/70682032/199153063-4c9c889e-664c-42f2-85e3-3858cf9fdc98.png">

### `doom-one-light`
<img width="769" alt="image" src="https://user-images.githubusercontent.com/70682032/199153112-36d8d89f-9c2b-4d75-9f8b-8c5efeb3ade8.png">

Here's a portion of the file used for the images:
```agda
module demo where

open import Relation.Binary.PropositionalEquality
open import Data.Nat

data List (A : Set) : Set where
  empty : List A
  add   : A → List A → List A

-- The list 1, 3, 2

ex1 : List ℕ
ex1 = add 1 (add 3 (add 2 empty))

append : {A : Set} → List A → List A → List A
append empty ys = ys
append (add x xs) ys = add x (append xs ys)

-- A test of append

ex2 : append (add 1 (add 5 (add 3 empty)))
             (add 2 (add 4 empty))
      ≡
      (add 1 (add 5 (add 3 (add 2 (add 4 empty)))))
ex2 = refl

append-empty-left : {A : Set} → forall (xs : List A) → append empty xs ≡ xs
append-empty-left xs = refl
```

**Comment by protesilaos** [comment]
Hello @hydrobeam!  This is because agda is not supported by the themes.  I will now look into it and report back to you.

**Comment by protesilaos** [comment]
Hello again!  Just pushed the change.

```
commit f1149c18005d31638b3701f9a89819b133b4360e
Author: Protesilaos Stavrou <info@protesilaos.com>
Date:   Tue Nov 1 06:24:38 2022 +0200

    Add support for agda2-mode
    
    This is based on merge request 44 by Xinglu Chen on the GitLab mirror
    from 2021-06-16:
    <https://gitlab.com/protesilaos/modus-themes/-/merge_requests/44>.
    That contribution could not be merged, as it exceeded the ~15 line
    limit and thus required copyright assignment to the Free Software
    Foundation.  The author became inactive soon thereafter despite their
    stated intention to do the paperwork.
    
    Thanks to Laith Bahodi for informing me about this in issue 48 on the
    GitHub mirror: <https://github.com/protesilaos/modus-themes/issues/48>.
    
    In principle, the agda project should be inheriting from font-lock
    faces instead of hardcoding their own colour values: those do not play
    nicely with themes.  If anyone familiar with agda reads this, I
    encourage you to work with upstream to make things better for all
    themes.  Also consider that adga cannot be tested without an
    established setup for Haskell, which adds to the burden of maintaining
    those faces at the theme level (I could not test them and only went by
    screenshots and prior experience).

 doc/modus-themes.info | 103 +++++++++++++++++++++++++-------------------------
 doc/modus-themes.org  |  31 +++++++--------
 modus-themes.el       |  31 +++++++++++++++
 3 files changed, 99 insertions(+), 66 deletions(-)
```

Please give it a try and let me know if it works for you.

**Comment by hydrobeam** [comment]
Thank you for your swift response and for the commit. I tried it out on the dev branch and the syntax highlighting does take effect! (I did have some troubles with upgrading my modus-themes version that I discuss after): 

`modus-vivendi`

<img width="739" alt="image" src="https://user-images.githubusercontent.com/70682032/199187948-fb957b1c-1ca6-4b33-ae40-7986d9c92565.png">

`modus-operandi`

<img width="676" alt="image" src="https://user-images.githubusercontent.com/70682032/199187989-a03b8532-d4b3-4d4b-aef2-547aafe3003c.png">

On an unrelated note, while trying to upgrade my modus version, I was running into errors at startup (unable to view anything, blank screen) that weren't present with other themes. So, I bisected the repo to try to find which commit might have introduced a bug, and narrowed it down to a commit between versions `2.2.0` and `2.3.0`:  

1706af0643f44dfe3879598934132c8f48aa113f

I'm not sure if there was an intended breaking change or if I was doing something improperly, but I checked the changelog and it seemed to not mention breaking changes.

https://protesilaos.com/emacs/modus-themes-changelog

I was eventually able to fix the startup issue by removing the following line from my config:

```lisp
(setq modus-themes-mode-line '(accented borderless (padding . 0) (height . 0.0)))
```

Here are the traces from using `emacs --debug-init`: (note there are two, one from the latest branch, and one from the commit that introduced the regression).

on `3.1.0-dev`:

```
Debugger entered--Lisp error: (error "Face height does not produce a positive integer" 0.0)
  set-face-attribute(mode-line-inactive #<frame *scratch* – Doom Emacs 0x5632925fe2b8> :inherit modus-themes-ui-variable-pitch :foreground "#bfc0c4" :background "#1e1e1e" :height 0.0 :box (:line-width 0 :color "#1e1e1e") :overline unspecified :underline unspecified :distant-foreground unspecified)
  face-spec-set-2(mode-line-inactive #<frame *scratch* – Doom Emacs 0x5632925fe2b8> (:inherit modus-themes-ui-variable-pitch :foreground "#bfc0c4" :background "#1e1e1e" :height 0.0 :box (:line-width 0 :color "#1e1e1e") :overline unspecified :underline unspecified :distant-foreground unspecified))
  face-spec-recalc(mode-line-inactive #<frame *scratch* – Doom Emacs 0x5632925fe2b8>)
  custom-theme-recalc-face(mode-line-inactive)
  enable-theme(modus-vivendi)
  #<subr load-theme>(modus-vivendi t nil)
  apply(#<subr load-theme> (modus-vivendi t nil))
  #f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a)(modus-vivendi t nil)
  funcall(#f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a) modus-vivendi t nil)
  (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes 'none)) (doom-run-hooks 'doom-load-theme-hook))))
  (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes 'none)) (doom-run-hooks 'doom-load-theme-hook)))))
  (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes 'none)) (doom-run-hooks 'doom-load-theme-hook))))))
  (unwind-protect (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes ...)) (doom-run-hooks 'doom-load-theme-hook)))))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))
  (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put ... ... ...) (doom-run-hooks ...)))))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))
  (let ((temp-buffer (generate-new-buffer " *temp*" t))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and ... ...) (progn ... ... ...))))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))
  doom--load-theme-a(#f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a) modus-vivendi t)
  apply(doom--load-theme-a #f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a) (modus-vivendi t))
  load-theme(modus-vivendi t)
  (progn (load-theme doom-theme t))
  (if (and doom-theme (not (custom-theme-enabled-p doom-theme))) (progn (load-theme doom-theme t)))
  doom-init-theme-h()
  run-hooks(after-init-hook delayed-warnings-hook)
  command-line()
  normal-top-level()
```

on 1706af0643f44dfe3879598934132c8f48aa113f
```
Debugger entered--Lisp error: (error "Invalid face box" :line-width 0 :color "#1e1e1e")
  set-face-attribute(mode-line-inactive #<frame *scratch* – Doom Emacs 0x55cc4f0ed1d8> :inherit modus-themes-ui-variable-pitch :foreground "#bfc0c4" :background "#1e1e1e" :box (:line-width 0 :color "#1e1e1e") :overline unspecified :underline unspecified :distant-foreground unspecified)
  face-spec-set-2(mode-line-inactive #<frame *scratch* – Doom Emacs 0x55cc4f0ed1d8> (:inherit modus-themes-ui-variable-pitch :foreground "#bfc0c4" :background "#1e1e1e" :box (:line-width 0 :color "#1e1e1e") :overline unspecified :underline unspecified :distant-foreground unspecified))
  face-spec-recalc(mode-line-inactive #<frame *scratch* – Doom Emacs 0x55cc4f0ed1d8>)
  custom-theme-recalc-face(mode-line-inactive)
  enable-theme(modus-vivendi)
  #<subr load-theme>(modus-vivendi t nil)
  apply(#<subr load-theme> (modus-vivendi t nil))
  #f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a)(modus-vivendi t nil)
  funcall(#f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a) modus-vivendi t nil)
  (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes 'none)) (doom-run-hooks 'doom-load-theme-hook))))
  (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes 'none)) (doom-run-hooks 'doom-load-theme-hook)))))
  (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes 'none)) (doom-run-hooks 'doom-load-theme-hook))))))
  (unwind-protect (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put 'doom-theme 'previous-themes (or last-themes ...)) (doom-run-hooks 'doom-load-theme-hook)))))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))
  (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and (not no-enable) (custom-theme-enabled-p theme)) (progn (setq doom-theme theme) (put ... ... ...) (doom-run-hooks ...)))))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))
  (let ((temp-buffer (generate-new-buffer " *temp*" t))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (let ((last-themes (copy-sequence custom-enabled-themes))) (mapc #'disable-theme custom-enabled-themes) (prog1 (funcall fn theme no-confirm no-enable) (if (and ... ...) (progn ... ... ...))))) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))
  doom--load-theme-a(#f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a) modus-vivendi t)
  apply(doom--load-theme-a #f(advice-wrapper :after #<subr load-theme> solaire-mode--prepare-for-theme-a) (modus-vivendi t))
  load-theme(modus-vivendi t)
  (progn (load-theme doom-theme t))
  (if (and doom-theme (not (custom-theme-enabled-p doom-theme))) (progn (load-theme doom-theme t)))
  doom-init-theme-h()
  run-hooks(after-init-hook delayed-warnings-hook)
  command-line()
  normal-top-level()
```

I am using Doom Emacs, if that may be relevant. I can split this off into a separate issue if you'd like too. 

Thanks again for adding agda support!

**Comment by protesilaos** [comment]
> Thanks again for adding agda support!

You are welcome!

> I was eventually able to fix the startup issue by removing the following line from my config: [...]

These values were not meant to be acceptable.  The `padding` expects a natural number, while the `height` must be a float that is not `0.0`.  The error handling can be improved here, though this is the intended behaviour.  I understand you want the mode lines to be just a thin line, right?

**Comment by hydrobeam** [comment]
> These values were not meant to be acceptable.
That's fair, I consulted the manual and the requirement for padding to be a natural number > 1 is mentioned. `height` not being `0.0` is not explicitly mentioned, but now that I understand what it does make sense why it's not valid.

>  I understand you want the mode lines to be just a thin line, right?

To be honest, I don't think I was thinking about what I wanted. I think I just set those values to experiment, saw that they didn't have an effect, and left it as such. The default size of the modeline is acceptable to me.
